### DISCUSSION TITLE
Forcing PySR to use specific formula and shortcut for activating all unary operators

### DISCUSSION THREAD

**Discussion by Sta1ger** [MilesCranmer/PySR#657] [Q&A] [answered]
Hi @MilesCranmer 
Frist of all thank you for that wonderful tool that you have created with PySR. Unfortunately I am fairly new to the data science world and am facing some difficulties as a result. I work in the field of material science and I’m measuring material properties at for different conditions for example temperatur. For my data analysis I have identified to steps, where I want to use or maybe misuse PySR.

In the first step I want to fit my data with PySR, but the mathematical function to fit the data is given by the literature as: A*( digmma(1/2-B/abs(x)) – ln(B/abs(x)) ) – C*abs(x)^2
Is there a way to force PySR to use a specific formula and only optimize the factors A, B and C (similar to functions like for example scipy.optimize.curve_fit)? I have already tried using different modules like scipy for this problem but often the optimisation fails.

In a second step I wanted to use PySR to find a temperature dependent expression for A, B and C.
As far as I know, we don’t know which kind of expression this temperature dependencies have. Therefore, is there a possible shortcut to activate all possible unary operators?

Thank you for your time and input in advance!

**Comment by MilesCranmer** [comment]
Hi @Sta1ger, since you are only optimising constants, perhaps you could use `scipy.optimize.minimize`? PySR is better for finding new expressions where you don't know the specific functional form

Related:
- https://github.com/milescranmer/pysr/discussions/631
- https://github.com/MilesCranmer/PySR/discussions/547

**Reply by Sta1ger** [reply]
Thank you for the fast reply @MilesCranmer, I have already applied various functions from scipy.optimize (curve_fit , minimize , basinhopping , differential_evolution) to this problem, but unfortunately the optimisation often fails. 

Could you elaborate on the second question, a shortcut to activate all unary operators?

**Reply by MilesCranmer** [reply, marked as answer]
I don’t recommend turning them all on because many are correlated with each other (sin and cos) and the more operators you have, the slower the search. Keep in mind that it’s roughly a O(n!) search space to explore, with `n` some function of the number of features and operators.
